File: java/src/test/java/br/com/vps/consulting/b2b/management/partner/domain/PartnerTest.java
package br.com.vps.consulting.b2b.management.partner.domain;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.time.Instant;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertThrows;

class PartnerTest {

    @Test
    @DisplayName("Given a provided id, when building a Partner, should create it with that id")
    void shouldCreateWithProvidedId() {
        final var id = PartnerId.generate();
        final var partner = newPartner(id);
        assertThat(partner.getId()).isEqualTo(id);
    }

    @Test
    @DisplayName("Given a null id, when building a Partner, should generate one")
    void shouldGenerateIdWhenNull() {
        final var partner = Partner.builder()
                .id(null)
                .name("Acme Corp")
                .document("12345678000100")
                .build();
        assertThat(partner.getId()).isNotNull();
    }

    @Test
    @DisplayName("Given a null createdAt, when building a Partner, should set the current time")
    void shouldSetCurrentTimeWhenCreatedAtIsNull() {
        final var before = Instant.now();
        final var partner = Partner.builder()
                .name("Acme Corp")
                .document("12345678000100")
                .createdAt(null)
                .build();
        assertThat(partner.getCreatedAt()).isAfterOrEqualTo(before);
    }

    @Test
    @DisplayName("Given a null name, when building a Partner, should reject it with NullPointerException")
    void shouldRejectNullName() {
        assertThrows(NullPointerException.class, () -> Partner.builder()
                .name(null)
                .document("12345678000100")
                .build());
    }

    @Test
    @DisplayName("Given a blank name, when building a Partner, should throw IllegalArgumentException")
    void shouldRejectBlankName() {
        assertThatThrownBy(() -> Partner.builder()
                .name("   ")
                .document("12345678000100")
                .build())
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessageContaining("não pode estar vazio");
    }

    @Test
    @DisplayName("Given a null document, when building a Partner, should reject it with NullPointerException")
    void shouldRejectNullDocument() {
        assertThrows(NullPointerException.class, () -> Partner.builder()
                .name("Acme Corp")
                .document(null)
                .build());
    }

    @Test
    @DisplayName("Given a blank document, when building a Partner, should throw IllegalArgumentException")
    void shouldRejectBlankDocument() {
        assertThatThrownBy(() -> Partner.builder()
                .name("Acme Corp")
                .document("   ")
                .build())
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessageContaining("não pode estar vazio");
    }

    private static Partner newPartner() {
        return newPartner(PartnerId.generate());
    }

    private static Partner newPartner(PartnerId id) {
        return Partner.builder()
                .id(id)
                .name("Acme Corp")
                .document("12345678000100")
                .createdAt(Instant.now())
                .build();
    }
}

File: kotlin/src/test/kotlin/br/com/vps/consulting/b2b/management/partner/domain/PartnerTest.kt
package br.com.vps.consulting.b2b.management.partner.domain

import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test
import java.time.Instant

class PartnerTest {

    @Test
    fun `should create partner via create with generated id`() {
        val partner = Partner.create(name = "Acme Corp", document = "12345678000100")

        assertThat(partner.id).isNotNull()
        assertThat(partner.name).isEqualTo("Acme Corp")
        assertThat(partner.document).isEqualTo("12345678000100")
    }

    @Test
    fun `should create partner via create with current timestamp as createdAt`() {
        val before = Instant.now()

        val partner = Partner.create(name = "Acme Corp", document = "12345678000100")

        assertThat(partner.createdAt).isAfterOrEqualTo(before)
    }

    @Test
    fun `should reject blank name via create`() {
        assertThatThrownBy { Partner.create(name = "   ", document = "12345678000100") }
            .isInstanceOf(IllegalArgumentException::class.java)
            .hasMessageContaining("não pode estar em branco")
    }

    @Test
    fun `should reject blank document via create`() {
        assertThatThrownBy { Partner.create(name = "Acme Corp", document = "   ") }
            .isInstanceOf(IllegalArgumentException::class.java)
            .hasMessageContaining("não pode estar em branco")
    }

    @Test
    fun `should reconstruct partner via with preserving all fields`() {
        val id = PartnerId.generate()
        val createdAt = Instant.now()

        val partner = Partner.with(id = id, name = "Acme Corp", document = "12345678000100", createdAt = createdAt)

        assertThat(partner.id).isEqualTo(id)
        assertThat(partner.name).isEqualTo("Acme Corp")
        assertThat(partner.document).isEqualTo("12345678000100")
        assertThat(partner.createdAt).isEqualTo(createdAt)
    }

    @Test
    fun `should reject blank name via with`() {
        assertThatThrownBy {
            Partner.with(id = PartnerId.generate(), name = "  ", document = "12345678000100", createdAt = Instant.now())
        }
            .isInstanceOf(IllegalArgumentException::class.java)
            .hasMessageContaining("não pode estar em branco")
    }

    @Test
    fun `should reject blank document via with`() {
        assertThatThrownBy {
            Partner.with(id = PartnerId.generate(), name = "Acme Corp", document = "  ", createdAt = Instant.now())
        }
            .isInstanceOf(IllegalArgumentException::class.java)
            .hasMessageContaining("não pode estar em branco")
    }

}
